feat: add support for monospace in pushoverConfig

heliapb opened the pull request:
## Description

Add support for monospace in pushoverConfig in alertmanager config, from version 0.29.0 onwards https://github.com/prometheus/alertmanager/pull/4362

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification
<!-- How you tested it? How do you know it works? -->
Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```release-note
feat: add support for monospace in pushoverConfig 
```

heliapb commented:
> lgtm probably can wait till Alertmanager v0.29 release

hi @slashpai I think this can be merged now. thanks

slashpai commented on pkg/alertmanager/amcfg.go:
nit
```suggestion
		return errors.New("either monospace or html must be configured")
```

slashpai commented on pkg/apis/monitoring/v1beta1/alertmanager_config_types.go:
```diff
@@ -1006,6 +1006,10 @@ type PushoverConfig struct {
+	// monospace optional HTML/monospace formatting for the message, see https://pushover.net/api#html
+	// html and monospace formatting are mutually exclusive.
+	// +optional
+	Monospace bool `json:"monospace,omitempty"`
```
since optional make this as `*bool`?

I see for HTML as well its bool may be we should update that

cc: @simonpasquier

heliapb commented:
> Thanks for the PR! We need to drop invalid AlertmanagerConfigs (e.g. with both monospace and html being true) before we sanitize. It should happen in checkPushoverConfigs().

thanks I added that and also noticed that I've missed the checks in validations, added those as well.

nutmos commented on pkg/apis/monitoring/v1beta1/alertmanager_config_types.go:
```diff
@@ -1005,7 +1005,11 @@ type PushoverConfig struct {
-	HTML bool `json:"html,omitempty"`
+	HTML *bool `json:"html,omitempty"`
+	// monospace optional HTML/monospace formatting for the message, see https://pushover.net/api#html
+	// html and monospace formatting are mutually exclusive.
```
html and monospace formatting are mutually exclusive.

Should this message be added to the HTML description as well?

simonpasquier replied:
I've merged to have the PR in the next release but the comment is valid and we can address in a follow-up PR.

simonpasquier commented on pkg/alertmanager/amcfg.go:
```diff
@@ -1283,7 +1283,6 @@ func (cb *ConfigBuilder) convertPushoverConfig(ctx context.Context, in monitorin
 		URL:           in.URL,
 		URLTitle:      in.URLTitle,
 		Priority:      in.Priority,
-		HTML:          in.HTML,
```
oops
